onset_maproom/calc.py

```diff
@@ -53,6 +332,228 @@ def water_balance(
+    reduced (or actual) evapotranspiration,
+    drain
+    """
+    # Start water balance ds with runoff
+    if runoff is None:
+        runoff = xr.zeros_like(daily_rain).rename("runoff")
+        runoff.attrs = dict(description="Runoff", units="mm")
+    water_balance = xr.Dataset().merge(runoff)
+    # Compute Effective Precipitation
+    peffective = (daily_rain - water_balance.runoff).rename("peffective")
+    peffective.attrs = dict(description="Effective Precipitation", units="mm")
+    water_balance = water_balance.merge(peffective)
+    # Additional vars and coords
+    et = (et * xr.ones_like(water_balance.peffective)).rename("et")
+    water_balance = water_balance.merge(et)
+    if kc_params is not None:
+        for adim in kc_params.dims:
+            for thedims in water_balance.dims:
+                if adim != thedims and adim != "kc_periods":
+                    water_balance[adim] = kc_params[adim]
+    if p_d is not None:
+        for adim in p_d.dims:
+            for thedims in water_balance.dims:
+                if adim != thedims and adim != time_coord:
+                    water_balance[adim] = p_d[adim]
+    # Get time_coord info
+    time_coord_size = water_balance.peffective[time_coord].size
+    # Intializing sm, et_crop and et_crop_red
+    # Creating variables and broadcasting everybody
+    soil_moisture = xr.full_like(
+        water_balance.peffective, np.nan
+    ).rename("soil_moisture")
+    soil_moisture.attrs = dict(description="Soil Moisture", units="mm")
+    water_balance = water_balance.merge(soil_moisture)
+    et_crop = xr.full_like(
+        water_balance.peffective, np.nan
+    ).rename("et_crop")
+    et_crop.attrs = dict(description="Crop Evapotranspiration", units="mm")
+    water_balance = water_balance.merge(et_crop)
+    (water_balance,) = xr.broadcast(water_balance)
+    water_balance["soil_moisture"] = water_balance.soil_moisture.copy()
+    water_balance["et_crop"] = water_balance.et_crop.copy()
+    # Give time dimension to sminit
+    sminit0 = xr.full_like(
+        water_balance.soil_moisture
+        .isel({time_coord: 0}).expand_dims(dim=time_coord),
+        sminit,
+    )
+    sminit0 = sminit0.assign_coords(
+        {time_coord: sminit0[time_coord] - np.timedelta64(1, "D")}
+    )
+    # Initialize Ks
+    ks = 1
+    if rho is not None:
+        raw = calibrate_available_water(taw, rho)
+        ks = single_stress_coeff(
+            sminit0,
+            taw,
+            raw.isel({time_coord: 0}, missing_dims="ignore"),
+            time_coord=time_coord,
+        ).squeeze(time_coord)
+    # Create or Initialize Kc
+    if kc_params is None:
+        kc0 = xr.ones_like(water_balance.et)
+    else:
+        if p_d is not None:
+            kc0 = kc_interpolation(p_d, kc_params, time_coord=time_coord)
+        else:
+            p_d_find = planting_date(
+                sminit0, sm_threshold, time_coord=time_coord
+            ).expand_dims(dim=time_coord)
```

Comment:
I think this relates to point 9 in your description:

> et_crop is built/saved inside the loop, whereas et_crop_red and drain, outside the loop. It could be either way for the 3 variables. Any sense if a way should be better?

I think you're actually calculating et_crop_red and drain twice each, once inside the loop for use in the stepwise calculation and then again at the end for saving into the result. The second time seems redundant, which makes the program both slower and more likely to be wrong (if we modify the calculation in one place and forget to modify the other).

Reply:
For planting_date and kc_interpolation, I believe your comment is correct. The current end result looks like this because I tried to reconcile:
1/ one loop fit all (which I thought was a goal at some point);
2/ modularity by having planting_date and kc_interpolation be "stand alone" functions that could be replaced by another.
As of now, the only other option I can think of is to have cases trigger different loops. Some cases will have planting_date and kc outside the loop, and some outside. Those inside the loop will be their own function, those outside will be the vectorized version. The case of planting_date is less trivial because it can vary in the other dimensions so you can have null in some locations and non-null (thus "savable" elsewhere).

The et_crop_red/drain is partially true. They are not computing per say in the loops but could trivially add subsets to "save" them there. I set up both cases (inside/outside the loop) as it's not clear to me what's better. I have a sense that the least in the loop, the better, but maybe that is not the right rule of thumbs.

What do you think of all that?

Reply:
If this still isn't clear, I can try expressing myself in code rather than in English.

Reply:
Regarding et_crop_red and drain, I'm not sure I understand your point of view. Yes, if you could calculate it outside of the loop, and not inside, that would be better, but that's not an option, because each step's soil moisture depends on the previous step's ET and drainage. The best feasible option is to do it inside the loop. Doing it inside the loop and then again after the loop exits is clearly worse than that. Maybe I'm still missing something.

Reply:
ok. So we can have planting_date, when estimated, being "saved" step after step of the loop; got back to previous step rather than origin of times; use where or something to search for the new time step where it's not found yet. That seems workable. Maybe only because planting_date depends only on conditions in one day (as opposed to onset_date for instance the looks at at n-day window). But that is probably going to be modul-able later on if such a scenario arise.

After I've done that I'll have a look at Kc in the same spirit.

For et_crop_red/drain, indeed, et_crop_red is literally computed in the loop by reduce_crop_evapotranspiration. Drainage is not as things stand (the clip filter destroys the Drainage information) but it's a trivial matter to add to the loop to save it before clipping. So that is what I meant, that in that very case, Drainage is not computed inside the loop (but close). I don't think you are missing something, I think it's me who is being extremely literal in trying to keep things outside the loop.

So I'll work on all that now (ish) unless you raise a flag reading that conversation.

Reply:
No red flag.

Reply:
With the latest changes, `test_soil_plant_water_balance_with_et_crop_pd_none` has gone from 1.6s to 1s, which is a significant speedup but stll slower than I would expect given the size of the data.

Do you anticipate that this function will be used in maprooms, or only in offline/batch scripts? If in maprooms, I think we need to continue working on performance (I can help). If it's only going to be run in batches, maybe it doesn't matter.

Reply:
Historically, the initial desire was to run the simulation via a Maproom that controls some of the parametric inputs; but in Ingrid, it wasn't possible (too long to run) so we ran several simulations of interest once and then Maprooms crunched the daily outputs of that ([e.g.](http://maproom.meteorwanda.gov.rw/maproom/Agriculture/index.html#tabs-2) each of these has a different Kc+planting). I think on a short-term, PIs can be amenable to similar set up. In longer term, they might push for being able to run the simulation via a Maproom.

There was also always a desire to have Monitoring Maprooms using such simulation, but it never happened because there never was real-time temperature data. On short-term, that is likely going to remain so, but on mid-term, they might start producing those real-time ENACTS temperature. But the performance issue is a lesser problem then because you'd want to run like 1 year of simulation. Unless PIs want to compare real-time with long-term (e.g. anomalies), in which case, we are back to first paragraph...

So in short, my best guess is that performance issues could be addressed in a later PR, given Jim said this is going to be a deliverable for AICCRA but not ACToday; and I can get a better sense of what functionality he seeks in the meantime. Unless continuing to work on performance is going to shake the code quite a bit, in which case maybe it's better to do it now?

Reply:
It's easier to work on performance when there's good test coverage (so you can verify that your performance changes haven't broken it), so let's not worry about it for now.